pkg/server/tls.go

```diff
@@ -0,0 +1,85 @@
+package server
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"io/ioutil"
+	"sync"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+var (
+	clientCertExpiry = newLazyGauge(prometheus.GaugeOpts{
+		Name: "grpc_client_tls_certificate_expiry_seconds",
+		Help: "Expiration time of the gRPC client TLS certificate in seconds since the Unix epoch.",
+	})
+	serverCertExpiry = newLazyGauge(prometheus.GaugeOpts{
+		Name: "grpc_server_tls_certificate_expiry_seconds",
+		Help: "Expiration time of the gRPC server TLS certificate in seconds since the Unix epoch.",
+	})
+)
+
+type lazyGauge struct {
+	init   sync.Once
+	metric prometheus.Gauge
+}
+
+func newLazyGauge(opts prometheus.GaugeOpts) *lazyGauge {
+	return &lazyGauge{metric: prometheus.NewGauge(opts)}
+}
+
+func (g *lazyGauge) Set(v float64) {
+	g.init.Do(func() { prometheus.MustRegister(g.metric) })
+	g.metric.Set(v)
+}
+
+func loadVerifiedCerts(certFile, keyFile, caFile string, usage x509.ExtKeyUsage) (cert tls.Certificate, caPool *x509.CertPool, expiry time.Time, err error) {
+	cert, caPool, err = loadCerts(certFile, keyFile, caFile)
+	if err != nil {
+		return tls.Certificate{}, nil, time.Time{}, err
+	}
+	chains, err := verify(cert, caPool, usage)
+	if err != nil {
+		return tls.Certificate{}, nil, time.Time{}, err
+	}
+	return cert, caPool, earliestExpiry(chains), nil
+}
+
+func loadCerts(certFile, keyFile, caFile string) (tls.Certificate, *x509.CertPool, error) {
+	cert, err := tls.LoadX509KeyPair(certFile, keyFile)
+	if err != nil {
+		return tls.Certificate{}, nil, err
+	}
+	ca, err := ioutil.ReadFile(caFile)
+	if err != nil {
+		return tls.Certificate{}, nil, err
+	}
```

Comment:
we've lost some detail in the returned error here, previously we specified things like `error loading keypair` in the returned error message, which can just make debugging it a little easier for people.

Reply:
Done.